File: study/lua/src/app/data/ActivityWeeklyGiftPkgData.lua
--豪华礼包活动数据类
--@Author:Conley
local ActivityBaseData = import(".ActivityBaseData")
local ActivityConst = require("app.const.ActivityConst")
local BaseData = require("app.data.BaseData")
local LogicCheckHelper  = require("app.utils.LogicCheckHelper")
local ActWeekDiscount = require("app.config.act_week_discount")
---------------------------------------------------
local ActivityWeeklyGiftPkgUnitData = class("ActivityWeeklyGiftPkgUnitData", BaseData)

local schema = {}
schema["time"] 		= {"number", 0}
schema["id"] 		= {"number",0}
schema["config"] 	= {"table",{}}
ActivityWeeklyGiftPkgUnitData.schema = schema

function ActivityWeeklyGiftPkgUnitData:ctor(properties)
	ActivityWeeklyGiftPkgUnitData.super.ctor(self, properties)
end

-- 清除
function ActivityWeeklyGiftPkgUnitData:clear()
end

-- 重置
function ActivityWeeklyGiftPkgUnitData:reset()
end

function ActivityWeeklyGiftPkgUnitData:initData(id,time)
	self:setTime(time)
	self:setId(id)

	local info = ActWeekDiscount.get(id)
	assert(info,"act_week_discount not find id "..tostring(id))

	self:setConfig(info)
end

function ActivityWeeklyGiftPkgUnitData:getRemainBuyTime()
	local cfg = self:getConfig()
	return cfg.time - self:getTime()
end


--检查是否能购买
function ActivityWeeklyGiftPkgUnitData:checkIsCanBuy()
	local cfg = self:getConfig()
	--元宝和VIP条件
	local vip = cfg.vip
	local gold = cfg.price
	local checkParams = {
			[1] = { funcName = "enoughVip", param =  {vip}  },  --检查资源是否满足
			[2] = { funcName = "enoughCash", param =  {gold}  },  --检查资源是否满足
	}
	local success, errorMsg,funcName  = LogicCheckHelper.doCheckList(checkParams)
	return success, errorMsg,funcName
end

function ActivityWeeklyGiftPkgUnitData:checkVip()
	local cfg = self:getConfig()
	local success, hintCallback = LogicCheckHelper.enoughVip(cfg.vip)
	return success
end


---------------------------------------------------
local ActivityWeeklyGiftPkgData = class("ActivityWeeklyGiftPkgData", BaseData)

local schema = {}
schema["baseActivityData"] 	= {"table", {}}
ActivityWeeklyGiftPkgData.schema = schema

function ActivityWeeklyGiftPkgData:ctor(properties)
	ActivityWeeklyGiftPkgData.super.ctor(self, properties)

	local TimeExpiredData = require("app.data.TimeExpiredData")
	self:setResetType(TimeExpiredData.RESET_TYPE_WEEKLY)

	local activityBaseData = ActivityBaseData.new()
	activityBaseData:initData({id = ActivityConst.ACT_ID_WEEKLY_GIFT_PKG  })
	self:setBaseActivityData(activityBaseData)

	self._unitDataList = {}

	self._s2cGetActDiscountListener = G_NetworkManager:add(MessageIDConst.ID_S2C_GetActDiscount, handler(self, self._s2cGetActDiscount))
	self._s2cActDiscountListener = G_NetworkManager:add(MessageIDConst.ID_S2C_ActDiscount, handler(self, self._s2cActDiscount))
end

-- 清除
function ActivityWeeklyGiftPkgData:clear()
	ActivityWeeklyGiftPkgData.super.clear(self)
	self._s2cGetActDiscountListener:remove() 
	self._s2cGetActDiscountListener = nil

	self._s2cActDiscountListener:remove() 
	self._s2cActDiscountListener = nil


	self:getBaseActivityData():clear()
end

-- 重置
function ActivityWeeklyGiftPkgData:reset()
	ActivityWeeklyGiftPkgData.super.reset(self)
	self:getBaseActivityData():reset()
end

function ActivityWeeklyGiftPkgData:_setUnitData(id,time)
	local unitData = ActivityWeeklyGiftPkgUnitData.new()
	unitData:initData(id,time)
	self._unitDataList["k_"..tostring(id)] =  unitData
end

function ActivityWeeklyGiftPkgData:getUnitData(id)
	return self._unitDataList["k_"..tostring(id)] 
end

function ActivityWeeklyGiftPkgData:getAllShowUnitDatas()
	local dataList = {}
	local myVip = G_UserData:getVip():getLevel()

	for k,v in pairs(self._unitDataList) do
		if myVip >= v:getConfig().vip_show then
			table.insert( dataList,v)
		end
	end
	local sortFunc = function(obj1,obj2)
		local time1 = obj1:getRemainBuyTime() 
		local time2 = obj2:getRemainBuyTime()
		if time1 <= 0 and time2 <= 0 then
			return obj1:getId() > obj2:getId()
		elseif time1 <= 0 then
			return false
		elseif time2 <= 0 then
			return true
		end
		return obj1:getId() < obj2:getId()
	end
	table.sort(dataList,sortFunc)
	return dataList
end

function ActivityWeeklyGiftPkgData:_s2cGetActDiscount(id,message)
	if message.ret ~= MessageErrorConst.RET_OK then
		return
	end
	if message.discount_type ~= ActivityConst.GIFT_PKG_TYPE_WEEKLY then
		return 
	end
	self:getBaseActivityData():setHasData(true)
	self:resetTime()

	self._unitDataList = {}

	local goodsIds = rawget(message,"goods_ids") or {}
	local buyTimes = rawget(message,"buy_times") or {}
	
	for k,v in ipairs(goodsIds) do
		if v and buyTimes[k] then
			self:_setUnitData(v,buyTimes[k])
		end
	end

	G_SignalManager:dispatch(SignalConst.EVENT_WELFARE_GIFT_PKG_GET_INFO,id,message)
	G_SignalManager:dispatch(SignalConst.EVENT_RED_POINT_UPDATE,FunctionConst.FUNC_WELFARE,{actId =ActivityConst.ACT_ID_WEEKLY_GIFT_PKG})
end

function ActivityWeeklyGiftPkgData:_s2cActDiscount(id,message)
	if message.ret ~= MessageErrorConst.RET_OK then
		return
	end
	if message.discount_type ~= ActivityConst.GIFT_PKG_TYPE_WEEKLY then
		return 
	end
	local ids = rawget(message,"id") or {} 
	local giftId = 0
	for k,id in ipairs(ids) do
		local unitData = self:getUnitData(id)
		if unitData then
			giftId = id
			unitData:setTime(unitData:getTime() + 1)
		end
	end
	G_SignalManager:dispatch(SignalConst.EVENT_WELFARE_GIFT_PKG_GET_REWARD,id,message,giftId)
	G_SignalManager:dispatch(SignalConst.EVENT_RED_POINT_UPDATE,FunctionConst.FUNC_WELFARE,{actId =ActivityConst.ACT_ID_WEEKLY_GIFT_PKG})
end

function ActivityWeeklyGiftPkgData:c2sActDiscount(id)
	if self:isExpired() == true then
		self:pullData()
		return 
	end
	G_NetworkManager:send(MessageIDConst.ID_C2S_ActDiscount,  {discount_type = ActivityConst.GIFT_PKG_TYPE_WEEKLY,id = { id}  })
end

function ActivityWeeklyGiftPkgData:c2sGetActDiscount(discountType)
	G_NetworkManager:send(MessageIDConst.ID_C2S_GetActDiscount, {discount_type = discountType})
end

function ActivityWeeklyGiftPkgData:pullData()

	self:c2sGetActDiscount(ActivityConst.GIFT_PKG_TYPE_WEEKLY)
end

function ActivityWeeklyGiftPkgData:resetData()
	self:pullData()
	self:setNotExpire()--避免重复取数据
end

function ActivityWeeklyGiftPkgData:hasRedPoint()
	local showed = G_UserData:getRedPoint():isTodayShowedRedPointByFuncId(
		FunctionConst.FUNC_WELFARE,{actId = ActivityConst.ACT_ID_WEEKLY_GIFT_PKG}
    )
	if showed then
		return false
	end

	return not self:hasBuyAllGoods()
end

function ActivityWeeklyGiftPkgData:hasBuyAllGoods()
	for k,weekUnitData in pairs(self._unitDataList) do
		local buyTime = weekUnitData:getTime()
		if buyTime <=  0 then
			return false
		end 
	end
	return true
end

function ActivityWeeklyGiftPkgData:getRewards(ids)
	local awards = {}
	for k,id in ipairs(ids) do
		local unitData = self:getUnitData(id)
		if unitData then
			local cfg = unitData:getConfig()
			table.insert(awards,{type = cfg.type,value = cfg.value,size = cfg.size})
		end
	end
	return awards
end

return ActivityWeeklyGiftPkgData

File: study/ts/script/data/ActivityWeeklyGiftPkgData.ts
import { BaseData } from './BaseData';
import { ActivityBaseData } from './ActivityBaseData';
import { G_ConfigLoader, G_NetworkManager, G_SignalManager, G_UserData } from '../init';
import { ConfigNameConst } from '../const/ConfigNameConst';
import { LogicCheckHelper } from '../utils/LogicCheckHelper';
import { TimeExpiredData } from './TimeExpiredData';
import { ActivityConst } from '../const/ActivityConst';
import { Slot } from '../utils/event/Slot';
import { MessageIDConst } from '../const/MessageIDConst';
import { MessageErrorConst } from '../const/MessageErrorConst';
import { SignalConst } from '../const/SignalConst';
import { FunctionConst } from '../const/FunctionConst';
let unitSchema = {};
unitSchema['time'] = [
    'number',
    0
];
unitSchema['id'] = [
    'number',
    0
];
unitSchema['config'] = [
    'object',
    {}
];
let schema = {};
schema['baseActivityData'] = [
    'object',
    {}
];
export interface ActivityWeeklyGiftPkgUnitData {
getTime(): number
setTime(value: number): void
getLastTime(): number
getId(): number
setId(value: number): void
getLastId(): number
getConfig(): any
setConfig(value: any): void
getLastConfig(): any
}
export class ActivityWeeklyGiftPkgUnitData extends BaseData {
    public static schema = unitSchema;

    constructor (properties?) {
        super(properties);
    }
    public clear () {
    }
    public reset () {
    }
    public initData (id, time) {
        this.setTime(time);
        this.setId(id);
        let ActWeekDiscount = G_ConfigLoader.getConfig(ConfigNameConst.ACT_WEEK_DISCOUNT);
        let info = ActWeekDiscount.get(id);
        console.assert(info, 'act_week_discount not find id ' + String(id));
        this.setConfig(info);
    }
    public getRemainBuyTime () {
        let cfg = this.getConfig();
        return cfg.time - this.getTime();
    }
    public checkIsCanBuy () {
        let cfg = this.getConfig();
        let vip = cfg.vip;
        let gold = cfg.price;
        let checkParams = {
            [1]: {
                funcName: 'enoughVip',
                param: [vip]
            },
            [2]: {
                funcName: 'enoughCash',
                param: [gold]
            }
        };
        let [success, errorMsg, funcName] = LogicCheckHelper.doCheckList(checkParams);
        return [
            success,
            errorMsg,
            funcName
        ];
    }
    public checkVip () {
        let cfg = this.getConfig();
        let [success, hintCallback] = LogicCheckHelper.enoughVip(cfg.vip);
        return success;
    }
}
export interface ActivityWeeklyGiftPkgData {
getBaseActivityData(): ActivityBaseData
setBaseActivityData(value: ActivityBaseData): void
getLastBaseActivityData(): ActivityBaseData
getExpiredTime():number
}
export class ActivityWeeklyGiftPkgData extends BaseData {
    public static schema = schema;

        public _unitDataList;
        public _s2cGetActDiscountListener: Slot;
        public _s2cActDiscountListener: Slot;

    constructor (properties?) {
        super(properties);
        this.setResetType(TimeExpiredData.RESET_TYPE_WEEKLY);
        let activityBaseData = new ActivityBaseData();
        activityBaseData.initData({ id: ActivityConst.ACT_ID_WEEKLY_GIFT_PKG });
        this.setBaseActivityData(activityBaseData);
        this._unitDataList = {};
        this._s2cGetActDiscountListener = G_NetworkManager.add(MessageIDConst.ID_S2C_GetActDiscount, this._s2cGetActDiscount.bind(this));
        this._s2cActDiscountListener = G_NetworkManager.add(MessageIDConst.ID_S2C_ActDiscount, this._s2cActDiscount.bind(this));
    }
    public clear () {
        super.clear();
        this._s2cGetActDiscountListener.remove();
        this._s2cGetActDiscountListener = null;
        this._s2cActDiscountListener.remove();
        this._s2cActDiscountListener = null;
        this.getBaseActivityData().clear();
    }
    public reset () {
        super.reset();
        this.getBaseActivityData().reset();
    }
    public _setUnitData (id, time) {
        let unitData = new ActivityWeeklyGiftPkgUnitData();
        unitData.initData(id, time);
        this._unitDataList['k_' + String(id)] = unitData;
    }
    public getUnitData (id) {
        return this._unitDataList['k_' + String(id)];
    }
    public getAllShowUnitDatas () {
        let dataList = [];
        let myVip = G_UserData.getVip().getLevel();

        for (let k in this._unitDataList) {
            let v = this._unitDataList[k];
            if (myVip >= v.getConfig().vip_show)
                dataList.push(v);
        }
        let sortFunc = function (obj1, obj2) {
            let time1 = obj1.getRemainBuyTime();
            let time2 = obj2.getRemainBuyTime();
            if (time1 <= 0 && time2 <= 0) {
                return obj2.getId() - obj1.getId();
            } else if (time1 <= 0) {
                return 1;
            } else if (time2 <= 0) {
                return -1;
            }
            return obj1.getId() - obj2.getId();
        };
        dataList.sort(sortFunc);
        return dataList;
    }
    public _s2cGetActDiscount (id, message) {
        if (message.ret != MessageErrorConst.RET_OK) {
            return;
        }
        if (message.discount_type != ActivityConst.GIFT_PKG_TYPE_WEEKLY) {
            return;
        }
        this.getBaseActivityData().setHasData(true);
        this.resetTime();
        this._unitDataList = {};
        let goodsIds:any[] = message['goods_ids'] || [];
        let buyTimes:any[] = message['buy_times'] || [];
        for (let k=0; k<goodsIds.length; k++) {
            let v = goodsIds[k];
            if (k < buyTimes.length) {
                this._setUnitData(v, buyTimes[k]);
            }
        }
        G_SignalManager.dispatch(SignalConst.EVENT_WELFARE_GIFT_PKG_GET_INFO, id, message);
        G_SignalManager.dispatch(SignalConst.EVENT_RED_POINT_UPDATE, FunctionConst.FUNC_WELFARE, { actId: ActivityConst.ACT_ID_WEEKLY_GIFT_PKG });
    }
    public _s2cActDiscount (id, message) {
        if (message.ret != MessageErrorConst.RET_OK) {
            return;
        }
        if (message.discount_type != ActivityConst.GIFT_PKG_TYPE_WEEKLY) {
            return;
        }
        let ids = message['id'] || {};
        let giftId = 0;
        for (let k in ids) {
            let id = ids[k];
            let unitData = this.getUnitData(id);
            if (unitData) {
                giftId = id;
                unitData.setTime(unitData.getTime() + 1);
            }
        }
        G_SignalManager.dispatch(SignalConst.EVENT_WELFARE_GIFT_PKG_GET_REWARD, id, message, giftId);
        G_SignalManager.dispatch(SignalConst.EVENT_RED_POINT_UPDATE, FunctionConst.FUNC_WELFARE, { actId: ActivityConst.ACT_ID_WEEKLY_GIFT_PKG });
    }
    public c2sActDiscount (id) {
        if (this.isExpired() == true) {
            this.pullData();
            return;
        }
        G_NetworkManager.send(MessageIDConst.ID_C2S_ActDiscount, {
            discount_type: ActivityConst.GIFT_PKG_TYPE_WEEKLY,
            id: [id]
        });
    }
    public c2sGetActDiscount (discountType) {
        G_NetworkManager.send(MessageIDConst.ID_C2S_GetActDiscount, { discount_type: discountType });
    }
    public pullData () {
        this.c2sGetActDiscount(ActivityConst.GIFT_PKG_TYPE_WEEKLY);
    }
    public resetData () {
        this.pullData();
        this.setNotExpire();
    }
    public hasRedPoint () {
        let showed = G_UserData.getRedPoint().isTodayShowedRedPointByFuncId(FunctionConst.FUNC_WELFARE, { actId: ActivityConst.ACT_ID_WEEKLY_GIFT_PKG });
        if (showed) {
            return false;
        }
        return !this.hasBuyAllGoods();
    }
    public hasBuyAllGoods () {
        for (let k in this._unitDataList) {
            let weekUnitData = this._unitDataList[k];
            let buyTime = weekUnitData.getTime();
            if (buyTime <= 0) {
                return false;
            }
        }
        return true;
    }
    public getRewards (ids) {
        let awards = [];
        for (let k in ids) {
            let id = ids[k];
            let unitData = this.getUnitData(id);
            if (unitData) {
                let cfg = unitData.getConfig();
                awards.push({
                    type: cfg.type,
                    value: cfg.value,
                    size: cfg.size
                });
            }
        }
        return awards;
    }
}